File: br_perception/src/lidar/lidar_preprocessor.cpp
#include "br_perception/lidar/lidar_preprocessor.hpp"

#include <pcl/filters/approximate_voxel_grid.h>
#include <pcl/filters/statistical_outlier_removal.h>
#include <pcl/filters/crop_box.h>
#include <pcl_conversions/pcl_conversions.h>

#include <chrono>
#include <string>
#include <algorithm>

namespace br_perception {
namespace lidar {

// ============================================================================
// 构造 & 参数初始化
// ============================================================================
LidarPreprocessor::LidarPreprocessor(const rclcpp::NodeOptions& options)
  : Node("lidar_preprocessor", options)
{
  // 声明所有参数
  this->declare_parameter<double>("leaf_size", leaf_size_);
  this->declare_parameter<bool>("enable_outlier_filter", enable_outlier_filter_);
  this->declare_parameter<int>("outlier_mean_k", outlier_mean_k_);
  this->declare_parameter<double>("outlier_std_thresh", outlier_std_thresh_);
  this->declare_parameter<double>("roi_min_x", roi_min_x_);
  this->declare_parameter<double>("roi_max_x", roi_max_x_);
  this->declare_parameter<double>("roi_min_y", roi_min_y_);
  this->declare_parameter<double>("roi_max_y", roi_max_y_);
  this->declare_parameter<double>("roi_min_z", roi_min_z_);
  this->declare_parameter<double>("roi_max_z", roi_max_z_);
  this->declare_parameter<bool>("enable_self_filter", enable_self_filter_);
  this->declare_parameter<double>("self_radius_x", self_radius_x_);
  this->declare_parameter<double>("self_radius_y", self_radius_y_);
  this->declare_parameter<double>("self_radius_z", self_radius_z_);

  // 读取参数
  this->get_parameter("leaf_size", leaf_size_);
  this->get_parameter("enable_outlier_filter", enable_outlier_filter_);
  this->get_parameter("outlier_mean_k", outlier_mean_k_);
  this->get_parameter("outlier_std_thresh", outlier_std_thresh_);
  this->get_parameter("roi_min_x", roi_min_x_);
  this->get_parameter("roi_max_x", roi_max_x_);
  this->get_parameter("roi_min_y", roi_min_y_);
  this->get_parameter("roi_max_y", roi_max_y_);
  this->get_parameter("roi_min_z", roi_min_z_);
  this->get_parameter("roi_max_z", roi_max_z_);
  this->get_parameter("enable_self_filter", enable_self_filter_);
  this->get_parameter("self_radius_x", self_radius_x_);
  this->get_parameter("self_radius_y", self_radius_y_);
  this->get_parameter("self_radius_z", self_radius_z_);

  // 参数动态更新回调
  param_callback_handle_ = this->add_on_set_parameters_callback(
    std::bind(&LidarPreprocessor::on_parameter_change, this, std::placeholders::_1));

  // 订阅原始点云 (best-effort QoS，不阻塞发送端)
  sub_raw_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
    "/livox/lidar/pointcloud",
    rclcpp::SensorDataQoS(),
    std::bind(&LidarPreprocessor::pointcloud_callback, this, std::placeholders::_1));

  // 发布滤波后点云
  pub_filtered_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(
    "/perception/lidar/filtered",
    rclcpp::SensorDataQoS());

  // 发布诊断信息
  pub_diagnostics_ = this->create_publisher<diagnostic_msgs::msg::DiagnosticArray>(
    "/perception/lidar/diagnostics", 10);

  // 自过滤：订阅机器人位姿
  if (enable_self_filter_) {
    sub_pose_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      "/robot_pose", 10,
      std::bind(&LidarPreprocessor::pose_callback, this, std::placeholders::_1));
    RCLCPP_INFO(this->get_logger(), "Self-filter ENABLED");
  }

  RCLCPP_INFO(this->get_logger(),
    "LidarPreprocessor ready | leaf=%.3fm | outlier=%s(k=%d,σ=%.1f) | "
    "roi_x=[%.1f,%.1f] roi_y=[%.1f,%.1f] roi_z=[%.1f,%.1f]",
    leaf_size_,
    enable_outlier_filter_ ? "on" : "off",
    outlier_mean_k_, outlier_std_thresh_,
    roi_min_x_, roi_max_x_,
    roi_min_y_, roi_max_y_,
    roi_min_z_, roi_max_z_);
}

// ============================================================================
// 点云回调
// ============================================================================
void LidarPreprocessor::pointcloud_callback(
    const sensor_msgs::msg::PointCloud2::SharedPtr msg)
{
  // ── 解包 ──
  PointCloudPtr cloud_in = std::make_shared<PointCloud>();
  pcl::fromROSMsg(*msg, *cloud_in);

  const size_t input_points = cloud_in->size();
  if (input_points == 0) {
    RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(),
                          5000, "Empty point cloud received");
    return;
  }

  // ── 滤波管线（每步独立计时） ──
  auto t0 = std::chrono::steady_clock::now();

  auto cloud_ds    = downsample(cloud_in);
  auto t1 = std::chrono::steady_clock::now();

  auto cloud_clean = remove_outliers(cloud_ds);
  auto t2 = std::chrono::steady_clock::now();

  auto cloud_crop  = crop_roi(cloud_clean);
  auto t3 = std::chrono::steady_clock::now();

  auto cloud_final = remove_self(cloud_crop);
  auto t4 = std::chrono::steady_clock::now();

  // ── 耗时统计 ──
  using ms = std::chrono::duration<float, std::milli>;
  double total_ms     = ms(t4 - t0).count();
  double downsample_ms= ms(t1 - t0).count();
  double outlier_ms   = ms(t2 - t1).count();
  double crop_ms      = ms(t3 - t2).count();
  double self_ms      = ms(t4 - t3).count();

  // ── 发布滤波结果 ──
  sensor_msgs::msg::PointCloud2 output_msg;
  pcl::toROSMsg(*cloud_final, output_msg);
  output_msg.header = msg->header;
  pub_filtered_->publish(output_msg);

  // ── 发布诊断 ──
  publish_diagnostics(total_ms, downsample_ms, outlier_ms,
                      crop_ms, self_ms, input_points, cloud_final->size());

  // 延迟超标告警
  if (total_ms > 20.0f) {
    RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
      "Preprocessing slow: %.1f ms (input=%zu, output=%zu)",
      total_ms, input_points, cloud_final->size());
  }
}

// ============================================================================
// 滤波管线
// ============================================================================
LidarPreprocessor::PointCloudPtr LidarPreprocessor::filter_cloud(const PointCloudPtr& input)
{
  auto out = downsample(input);
  out = remove_outliers(out);
  out = crop_roi(out);
  out = remove_self(out);
  return out;
}

// ── 1. 降采样 ──
LidarPreprocessor::PointCloudPtr LidarPreprocessor::downsample(const PointCloudPtr& input)
{
  auto output = std::make_shared<PointCloud>();
  pcl::ApproximateVoxelGrid<pcl::PointXYZ> voxel;
  voxel.setInputCloud(input);
  voxel.setLeafSize(leaf_size_, leaf_size_, leaf_size_);
  voxel.filter(*output);
  return output;
}

// ── 2. 离群点过滤 ──
LidarPreprocessor::PointCloudPtr LidarPreprocessor::remove_outliers(const PointCloudPtr& input)
{
  if (!enable_outlier_filter_ || input->size() < static_cast<size_t>(outlier_mean_k_)) {
    return input;
  }

  auto output = std::make_shared<PointCloud>();
  pcl::StatisticalOutlierRemoval<pcl::PointXYZ> sor;
  sor.setInputCloud(input);
  sor.setMeanK(outlier_mean_k_);
  sor.setStddevMulThresh(outlier_std_thresh_);
  sor.setNegative(false);  // false = 保留内点, 剔除离群点
  sor.filter(*output);
  return output;
}

// ── 3. ROI 裁剪（XYZ 联合 CropBox，替代分散的 PassThrough×2 + CropBox） ──
LidarPreprocessor::PointCloudPtr LidarPreprocessor::crop_roi(const PointCloudPtr& input)
{
  auto output = std::make_shared<PointCloud>();

  pcl::CropBox<pcl::PointXYZ> box;
  box.setInputCloud(input);
  box.setMin(Eigen::Vector4f(roi_min_x_, roi_min_y_, roi_min_z_, 1.0f));
  box.setMax(Eigen::Vector4f(roi_max_x_, roi_max_y_, roi_max_z_, 1.0f));
  box.setNegative(false);  // 保留内部
  box.filter(*output);
  return output;
}

// ── 4. 自过滤（线程安全） ──
LidarPreprocessor::PointCloudPtr LidarPreprocessor::remove_self(const PointCloudPtr& input)
{
  if (!enable_self_filter_) {
    return input;
  }

  // 读 current_pose_ 时加锁（pose_callback 在另一线程写入）
  geometry_msgs::msg::PoseStamped pose;
  {
    std::lock_guard<std::mutex> lock(pose_mutex_);
    if (!has_pose_) {
      // 还没收到过位姿，跳过自过滤
      return input;
    }
    pose = current_pose_;
  }

  auto output = std::make_shared<PointCloud>();

  pcl::CropBox<pcl::PointXYZ> self_box;
  self_box.setInputCloud(input);
  self_box.setMin(Eigen::Vector4f(
    pose.pose.position.x - self_radius_x_,
    pose.pose.position.y - self_radius_y_,
    pose.pose.position.z - self_radius_z_,
    1.0f));
  self_box.setMax(Eigen::Vector4f(
    pose.pose.position.x + self_radius_x_,
    pose.pose.position.y + self_radius_y_,
    pose.pose.position.z + self_radius_z_,
    1.0f));
  self_box.setNegative(true);   // 剔除内部
  self_box.filter(*output);
  return output;
}

// ============================================================================
// 诊断
// ============================================================================
void LidarPreprocessor::publish_diagnostics(
    double total_ms,
    double downsample_ms,
    double outlier_ms,
    double crop_ms,
    double self_ms,
    size_t input_points,
    size_t output_points)
{
  diagnostic_msgs::msg::DiagnosticArray diag_array;
  diag_array.header.stamp = this->now();

  diagnostic_msgs::msg::DiagnosticStatus status;
  status.name = "lidar_preprocessor";
  status.hardware_id = "mid360";

  // 总体健康判定
  if (total_ms < 15.0) {
    status.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
    status.message = "OK";
  } else if (total_ms < 25.0) {
    status.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
    status.message = "SLOW";
  } else {
    status.level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
    status.message = "VERY SLOW";
  }

  // 各步骤耗时 + 点数信息 (key-value 对)
  auto add_kv = [&](const std::string& key, double val) {
    diagnostic_msgs::msg::KeyValue kv;
    kv.key = key;
    kv.value = std::to_string(val);
    status.values.push_back(kv);
  };

  add_kv("total_ms",       total_ms);
  add_kv("downsample_ms",  downsample_ms);
  add_kv("outlier_ms",     outlier_ms);
  add_kv("crop_ms",        crop_ms);
  add_kv("self_ms",        self_ms);
  add_kv("input_points",   static_cast<double>(input_points));
  add_kv("output_points",  static_cast<double>(output_points));
  add_kv("reduction_ratio",
         (input_points > 0) ? (1.0 - static_cast<double>(output_points) / input_points) : 0.0);
  add_kv("leaf_size",      leaf_size_);

  diag_array.status.push_back(status);
  pub_diagnostics_->publish(diag_array);
}

// ============================================================================
// 动态参数更新
// ============================================================================
rcl_interfaces::msg::SetParametersResult LidarPreprocessor::on_parameter_change(
    const std::vector<rclcpp::Parameter>& params)
{
  rcl_interfaces::msg::SetParametersResult result;
  result.successful = true;   // 默认成功
  result.reason = "";

  for (const auto& param : params) {
    const std::string& name = param.get_name();

    try {
      if (name == "leaf_size") {
        double val = param.as_double();
        if (val <= 0.0 || val > 1.0) {
          result.successful = false;
          result.reason = "leaf_size must be in (0, 1.0]";
          break;
        }
        leaf_size_ = val;
      }
      else if (name == "enable_outlier_filter") {
        enable_outlier_filter_ = param.as_bool();
      }
      else if (name == "outlier_mean_k") {
        int val = param.as_int();
        if (val < 3) {
          result.successful = false;
          result.reason = "outlier_mean_k must be >= 3";
          break;
        }
        outlier_mean_k_ = val;
      }
      else if (name == "outlier_std_thresh") {
        double val = param.as_double();
        if (val <= 0.0) {
          result.successful = false;
          result.reason = "outlier_std_thresh must be > 0";
          break;
        }
        outlier_std_thresh_ = val;
      }
      else if (name == "roi_min_x")  { roi_min_x_ = param.as_double(); }
      else if (name == "roi_max_x")  { roi_max_x_ = param.as_double(); }
      else if (name == "roi_min_y")  { roi_min_y_ = param.as_double(); }
      else if (name == "roi_max_y")  { roi_max_y_ = param.as_double(); }
      else if (name == "roi_min_z")  { roi_min_z_ = param.as_double(); }
      else if (name == "roi_max_z")  { roi_max_z_ = param.as_double(); }
      else if (name == "enable_self_filter") {
        enable_self_filter_ = param.as_bool();
        if (enable_self_filter_ && !sub_pose_) {
          sub_pose_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "/robot_pose", 10,
            std::bind(&LidarPreprocessor::pose_callback, this, std::placeholders::_1));
        }
      }
      else if (name == "self_radius_x")  { self_radius_x_ = param.as_double(); }
      else if (name == "self_radius_y")  { self_radius_y_ = param.as_double(); }
      else if (name == "self_radius_z")  { self_radius_z_ = param.as_double(); }
      else {
        // 未知参数（可能属于其他节点），不标记失败，忽略即可
        RCLCPP_DEBUG(this->get_logger(), "Ignoring unknown param: %s", name.c_str());
      }
    }
    catch (const rclcpp::ParameterTypeException& e) {
      result.successful = false;
      result.reason = std::string("Type mismatch for ") + name + ": " + e.what();
      break;
    }
  }

  if (result.successful) {
    RCLCPP_INFO(this->get_logger(),
      "Params updated | leaf=%.3f outlier=%s(k=%d,σ=%.1f) self=%s",
      leaf_size_, enable_outlier_filter_ ? "on" : "off",
      outlier_mean_k_, outlier_std_thresh_,
      enable_self_filter_ ? "on" : "off");
  } else {
    RCLCPP_WARN(this->get_logger(), "Param update rejected: %s", result.reason.c_str());
  }

  return result;
}

// ============================================================================
// 位姿回调（自过滤用，在独立线程中调用）
// ============================================================================
void LidarPreprocessor::pose_callback(const geometry_msgs::msg::PoseStamped::SharedPtr pose_msg)
{
  std::lock_guard<std::mutex> lock(pose_mutex_);
  current_pose_ = *pose_msg;
  has_pose_ = true;
}

}  // namespace lidar
}  // namespace br_perception

// ROS2 组件注册
#include <rclcpp_components/register_node_macro.hpp>
RCLCPP_COMPONENTS_REGISTER_NODE(br_perception::lidar::LidarPreprocessor)

File: src/lidar/lidar_preprocessor.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from rclpy.parameter import Parameter
from rcl_interfaces.msg import SetParametersResult
from sensor_msgs.msg import PointCloud2
from geometry_msgs.msg import PoseStamped
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus, KeyValue
from sensor_msgs import point_cloud2
import pclpy
from pclpy import pcl
import numpy as np
import threading
import time
from datetime import datetime

class LidarPreprocessor(Node):
    def __init__(self, options=None):
        super().__init__('lidar_preprocessor', options=options)

        # 声明参数
        self.declare_parameter('leaf_size', 0.1)
        self.declare_parameter('enable_outlier_filter', True)
        self.declare_parameter('outlier_mean_k', 50)
        self.declare_parameter('outlier_std_thresh', 1.0)
        self.declare_parameter('roi_min_x', -50.0)
        self.declare_parameter('roi_max_x', 50.0)
        self.declare_parameter('roi_min_y', -50.0)
        self.declare_parameter('roi_max_y', 50.0)
        self.declare_parameter('roi_min_z', -5.0)
        self.declare_parameter('roi_max_z', 5.0)
        self.declare_parameter('enable_self_filter', False)
        self.declare_parameter('self_radius_x', 1.0)
        self.declare_parameter('self_radius_y', 1.0)
        self.declare_parameter('self_radius_z', 1.0)

        # 读取参数
        self.leaf_size = self.get_parameter('leaf_size').value
        self.enable_outlier_filter = self.get_parameter('enable_outlier_filter').value
        self.outlier_mean_k = self.get_parameter('outlier_mean_k').value
        self.outlier_std_thresh = self.get_parameter('outlier_std_thresh').value
        self.roi_min_x = self.get_parameter('roi_min_x').value
        self.roi_max_x = self.get_parameter('roi_max_x').value
        self.roi_min_y = self.get_parameter('roi_min_y').value
        self.roi_max_y = self.get_parameter('roi_max_y').value
        self.roi_min_z = self.get_parameter('roi_min_z').value
        self.roi_max_z = self.get_parameter('roi_max_z').value
        self.enable_self_filter = self.get_parameter('enable_self_filter').value
        self.self_radius_x = self.get_parameter('self_radius_x').value
        self.self_radius_y = self.get_parameter('self_radius_y').value
        self.self_radius_z = self.get_parameter('self_radius_z').value

        # 动态参数回调
        self.add_on_set_parameters_callback(self.on_parameter_change)

        # 订阅原始点云（传感器数据QoS）
        qos_sensor = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
            depth=5
        )
        self.sub_raw = self.create_subscription(
            PointCloud2,
            '/livox/lidar/pointcloud',
            self.pointcloud_callback,
            qos_sensor
        )

        # 发布滤波后点云
        self.pub_filtered = self.create_publisher(
            PointCloud2,
            '/perception/lidar/filtered',
            qos_sensor
        )

        # 发布诊断信息
        self.pub_diagnostics = self.create_publisher(
            DiagnosticArray,
            '/perception/lidar/diagnostics',
            10
        )

        # 自过滤：订阅机器人位姿
        self.current_pose = None
        self.has_pose = False
        self.pose_mutex = threading.Lock()
        if self.enable_self_filter:
            self.sub_pose = self.create_subscription(
                PoseStamped,
                '/robot_pose',
                10,
                self.pose_callback
            )
            self.get_logger().info("Self-filter ENABLED")

        self.get_logger().info(
            f"LidarPreprocessor ready | leaf={self.leaf_size:.3f}m | "
            f"outlier={'on' if self.enable_outlier_filter else 'off'}(k={self.outlier_mean_k},σ={self.outlier_std_thresh:.1f}) | "
            f"roi_x=[{self.roi_min_x:.1f},{self.roi_max_x:.1f}] "
            f"roi_y=[{self.roi_min_y:.1f},{self.roi_max_y:.1f}] "
            f"roi_z=[{self.roi_min_z:.1f},{self.roi_max_z:.1f}]"
        )

    # ========================================================================
    # 点云回调
    # ========================================================================
    def pointcloud_callback(self, msg: PointCloud2):
        # 解包ROS消息到pcl点云
        cloud_in = self.ros_to_pcl(msg)
        input_points = cloud_in.size()
        if input_points == 0:
            self.get_logger().warning("Empty point cloud received", throttle_duration_sec=5.0)
            return

        # 滤波管线计时
        t0 = time.perf_counter()

        cloud_ds = self.downsample(cloud_in)
        t1 = time.perf_counter()

        cloud_clean = self.remove_outliers(cloud_ds)
        t2 = time.perf_counter()

        cloud_crop = self.crop_roi(cloud_clean)
        t3 = time.perf_counter()

        cloud_final = self.remove_self(cloud_crop)
        t4 = time.perf_counter()

        # 耗时统计（毫秒）
        total_ms = (t4 - t0) * 1000.0
        downsample_ms = (t1 - t0) * 1000.0
        outlier_ms = (t2 - t1) * 1000.0
        crop_ms = (t3 - t2) * 1000.0
        self_ms = (t4 - t3) * 1000.0

        # 发布滤波结果
        output_msg = self.pcl_to_ros(cloud_final, msg.header)
        self.pub_filtered.publish(output_msg)

        # 发布诊断
        self.publish_diagnostics(
            total_ms, downsample_ms, outlier_ms, crop_ms, self_ms,
            input_points, cloud_final.size()
        )

        # 延迟超标告警
        if total_ms > 20.0:
            self.get_logger().warning(
                f"Preprocessing slow: {total_ms:.1f} ms (input={input_points}, output={cloud_final.size()})",
                throttle_duration_sec=2.0
            )

    # ========================================================================
    # 滤波管线（各步骤独立）
    # ========================================================================
    def downsample(self, cloud):
        """近似体素降采样"""
        output = pcl.PointCloud.PointXYZ()
        voxel = pcl.filters.ApproximateVoxelGrid.PointXYZ()
        voxel.setInputCloud(cloud)
        voxel.setLeafSize(self.leaf_size, self.leaf_size, self.leaf_size)
        voxel.filter(output)
        return output

    def remove_outliers(self, cloud):
        """统计离群点去除"""
        if not self.enable_outlier_filter or cloud.size() < self.outlier_mean_k:
            return cloud
        output = pcl.PointCloud.PointXYZ()
        sor = pcl.filters.StatisticalOutlierRemoval.PointXYZ()
        sor.setInputCloud(cloud)
        sor.setMeanK(self.outlier_mean_k)
        sor.setStddevMulThresh(self.outlier_std_thresh)
        sor.setNegative(False)   # 保留内点
        sor.filter(output)
        return output

    def crop_roi(self, cloud):
        """ROI裁剪（XYZ联合）"""
        output = pcl.PointCloud.PointXYZ()
        box = pcl.filters.CropBox.PointXYZ()
        box.setInputCloud(cloud)
        box.setMin(
            self.roi_min_x, self.roi_min_y, self.roi_min_z, 1.0
        )
        box.setMax(
            self.roi_max_x, self.roi_max_y, self.roi_max_z, 1.0
        )
        box.setNegative(False)   # 保留内部
        box.filter(output)
        return output

    def remove_self(self, cloud):
        """自过滤（剔除机器人本体点云）"""
        if not self.enable_self_filter:
            return cloud

        with self.pose_mutex:
            if not self.has_pose:
                return cloud
            pose = self.current_pose

        output = pcl.PointCloud.PointXYZ()
        box = pcl.filters.CropBox.PointXYZ()
        box.setInputCloud(cloud)
        box.setMin(
            pose.position.x - self.self_radius_x,
            pose.position.y - self.self_radius_y,
            pose.position.z - self.self_radius_z,
            1.0
        )
        box.setMax(
            pose.position.x + self.self_radius_x,
            pose.position.y + self.self_radius_y,
            pose.position.z + self.self_radius_z,
            1.0
        )
        box.setNegative(True)   # 剔除内部
        box.filter(output)
        return output

    # ========================================================================
    # 诊断发布
    # ========================================================================
    def publish_diagnostics(self, total_ms, downsample_ms, outlier_ms,
                            crop_ms, self_ms, input_points, output_points):
        diag_array = DiagnosticArray()
        diag_array.header.stamp = self.get_clock().now().to_msg()

        status = DiagnosticStatus()
        status.name = 'lidar_preprocessor'
        status.hardware_id = 'mid360'

        # 总体健康判定
        if total_ms < 15.0:
            status.level = DiagnosticStatus.OK
            status.message = 'OK'
        elif total_ms < 25.0:
            status.level = DiagnosticStatus.WARN
            status.message = 'SLOW'
        else:
            status.level = DiagnosticStatus.ERROR
            status.message = 'VERY SLOW'

        # 添加key-value
        def add_kv(key, value):
            kv = KeyValue()
            kv.key = key
            kv.value = str(value)
            status.values.append(kv)

        add_kv('total_ms', total_ms)
        add_kv('downsample_ms', downsample_ms)
        add_kv('outlier_ms', outlier_ms)
        add_kv('crop_ms', crop_ms)
        add_kv('self_ms', self_ms)
        add_kv('input_points', float(input_points))
        add_kv('output_points', float(output_points))
        if input_points > 0:
            reduction = 1.0 - output_points / input_points
            add_kv('reduction_ratio', reduction)
        else:
            add_kv('reduction_ratio', 0.0)
        add_kv('leaf_size', self.leaf_size)

        diag_array.status.append(status)
        self.pub_diagnostics.publish(diag_array)

    # ========================================================================
    # 动态参数更新
    # ========================================================================
    def on_parameter_change(self, params):
        result = SetParametersResult()
        result.successful = True
        result.reason = ''

        for param in params:
            name = param.name
            try:
                if name == 'leaf_size':
                    val = param.value
                    if val <= 0.0 or val > 1.0:
                        result.successful = False
                        result.reason = 'leaf_size must be in (0, 1.0]'
                        break
                    self.leaf_size = val
                elif name == 'enable_outlier_filter':
                    self.enable_outlier_filter = param.value
                elif name == 'outlier_mean_k':
                    val = param.value
                    if val < 3:
                        result.successful = False
                        result.reason = 'outlier_mean_k must be >= 3'
                        break
                    self.outlier_mean_k = val
                elif name == 'outlier_std_thresh':
                    val = param.value
                    if val <= 0.0:
                        result.successful = False
                        result.reason = 'outlier_std_thresh must be > 0'
                        break
                    self.outlier_std_thresh = val
                elif name == 'roi_min_x':
                    self.roi_min_x = param.value
                elif name == 'roi_max_x':
                    self.roi_max_x = param.value
                elif name == 'roi_min_y':
                    self.roi_min_y = param.value
                elif name == 'roi_max_y':
                    self.roi_max_y = param.value
                elif name == 'roi_min_z':
                    self.roi_min_z = param.value
                elif name == 'roi_max_z':
                    self.roi_max_z = param.value
                elif name == 'enable_self_filter':
                    self.enable_self_filter = param.value
                    if self.enable_self_filter and not hasattr(self, 'sub_pose'):
                        self.sub_pose = self.create_subscription(
                            PoseStamped,
                            '/robot_pose',
                            10,
                            self.pose_callback
                        )
                elif name == 'self_radius_x':
                    self.self_radius_x = param.value
                elif name == 'self_radius_y':
                    self.self_radius_y = param.value
                elif name == 'self_radius_z':
                    self.self_radius_z = param.value
                else:
                    self.get_logger().debug(f'Ignoring unknown param: {name}')
            except Exception as e:
                result.successful = False
                result.reason = f'Type mismatch for {name}: {str(e)}'
                break

        if result.successful:
            self.get_logger().info(
                f'Params updated | leaf={self.leaf_size:.3f} '
                f'outlier={"on" if self.enable_outlier_filter else "off"}(k={self.outlier_mean_k},σ={self.outlier_std_thresh:.1f}) '
                f'self={"on" if self.enable_self_filter else "off"}'
            )
        else:
            self.get_logger().warn(f'Param update rejected: {result.reason}')

        return result

    # ========================================================================
    # 位姿回调（自过滤用）
    # ========================================================================
    def pose_callback(self, msg: PoseStamped):
        with self.pose_mutex:
            self.current_pose = msg.pose
            self.has_pose = True

    # ========================================================================
    # ROS <-> PCL 转换辅助
    # ========================================================================
    @staticmethod
    def ros_to_pcl(ros_msg):
        """将sensor_msgs.PointCloud2转换为pcl.PointCloud.PointXYZ"""
        # 读取点云到numpy数组
        points = []
        for p in point_cloud2.read_points(ros_msg, field_names=('x', 'y', 'z'), skip_nans=True):
            points.append((p[0], p[1], p[2]))
        cloud = pcl.PointCloud.PointXYZ()
        if points:
            cloud.from_array(np.array(points, dtype=np.float32))
        return cloud

    @staticmethod
    def pcl_to_ros(cloud, header):
        """将pcl.PointCloud.PointXYZ转换为sensor_msgs.PointCloud2"""
        # 提取点云数据
        pts = cloud.xyz  # numpy array (N, 3)
        if pts.size == 0:
            # 空点云
            return point_cloud2.create_cloud_xyz32(header, np.zeros((0, 3), dtype=np.float32))
        # 创建消息
        return point_cloud2.create_cloud_xyz32(header, pts.astype(np.float32))


def main(args=None):
    rclpy.init(args=args)
    node = LidarPreprocessor()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()